Add resolution and examples for analzyer. (#5729)
istio/istio.io: content/en/docs/reference/config/analysis/ist0109/index.md

layout: analysis-message
---

This message occurs when Istio detects an overlap between
[virtual service](/docs/reference/config/networking/virtual-service)
resources that conflict with one another. For example, multiple virtual

> services defined to use the same hostname and attached to mesh gateway

Reviewer: ```suggestion
services are defined to use the same hostname and attached to mesh gateway
```

> will generate this message. Note that Istio supports merging of virtual

Reviewer: ```suggestion
will generate an error message. Note that Istio supports merging of virtual
```

services that are attached to the ingress gateways.

## Resolution

To resolve this issue, you can take one of the following actions:

* Merge the conflicting virtual services into a single resource.
* Make the hostnames unique across virtual services attached to mesh gateway.
* Scope the resource to a specific namespace by setting the `exportTo` field.

## Examples

> The following `VirtualServices`, `productpage` in namespaces `team1` and `team2`

Reviewer 2: I found this sentence confusing to read at first. Here's a stab at an alternative:
```
The `productpage` virtual service in namespace `team1` conflicts with the `productpage` 
virtual service in namespace `team2` because both of the following are true:
  * they are attached to the default mesh gateway (no custom gateway is specified)
  * They have the same host `productpage.default.svc.cluster.local` defined.
```

It may also be wise to have a different name for each virtual service, so it's clearer that the conflict isn't that they share a name (which is fine) but that they share a host.

are in conflict as they are attached to the default mesh gateway, and have the
same host `productpage.default.svc.cluster.local` defined.

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: productpage
  namespace: team-1
spec:
  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: productpage
  namespace: team-2
spec:
  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage.team-2.svc.cluster.local
---
{{< /text >}}

You can resolve this issue by scoping these `VirtualServices` to their respective namespaces like this:

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: productpage
  namespace: team-1
spec:
  exportTo:

>   - "."

Reviewer 2: @geeknoid Is there any way to have YAML text in our docs, with highlighted lines? 

If not, do you think it's a feasible FR? This is an example of where such a thing could be really useful.

Author: I was thinking about that, I know some Hugo themes support this, not sure if the one in Istio supports it.

Reviewer 3: No, we don't have support to highlight individual lines in a text block.  It could probably be done with some annotation, similar to the @@ syntax we support for links. Perhaps %% can indicate a span of text that should be rendered distinctively.

Anyway, not at the moment...

  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: productpage
  namespace: team-2
spec:
  exportTo:
  - "."
  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage.team-2.svc.cluster.local
---
{{< /text >}}

Note that setting the `exportTo` field to `.` restricts the scope to the

> namespace where the virtual service resource exists.

Reviewer 2: I'd move this up to where you explain the resolution. e.g. 

```
You can resolve this issue by setting the `exportTo` field to `.` so that each virtual service 
is scoped only to its own namespace. 
```

layout: analysis-message
---

This message occurs when Istio detects an overlap between
[virtual service](/docs/reference/config/networking/virtual-service)
resources that conflict with one another. For example, multiple virtual

> services defined to use the same hostname and attached to mesh gateway

Reviewer 3: ```suggestion
services defined to use the same hostname and attached to a mesh gateway
```

will generate an error message. Note that Istio supports merging of virtual
services that are attached to the ingress gateways.

## Resolution

To resolve this issue, you can take one of the following actions:

* Merge the conflicting virtual services into a single resource.

> * Make the hostnames unique across virtual services attached to mesh gateway.

Reviewer 3: ```suggestion
* Make the hostnames unique across virtual services attached to a mesh gateway.
```

* Scope the resource to a specific namespace by setting the `exportTo` field.

## Examples

The `productpage` virtual service in namespace `team1` conflicts with the
`custom` virtual service in `team2` namespace because both of the following
are true:

* They are attached to the default "mesh" gateway as no custom gateway is
specified.
* They both define the same host `productpage.default.svc.cluster.local`
defined.

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: productpage
  namespace: team-1
spec:
  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: custom
  namespace: team-2
spec:
  hosts:
  - productpage.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: productpage.team-2.svc.cluster.local
---
{{< /text >}}

> You can resolve this issue by setting the `exportTo` field is set to `.` so

Reviewer 3: ```suggestion
You can resolve this issue by setting the `exportTo` field to `.` so
```
